File: java/src/main/java/com/ywh/problem/leetcode/medium/LeetCode414.java
package com.ywh.problem.leetcode.medium;

/**
 * 第三大的数
 * [数组]
 *
 * @author ywh
 * @since 28/12/2019
 */
public class LeetCode414 {

    /**
     * Time: O(n), Space: O(1)
     *
     * @param nums
     * @return
     */
    public int thirdMax(int[] nums) {
        int first = Integer.MIN_VALUE, second = Integer.MIN_VALUE, third = Integer.MIN_VALUE;

        // 处理数组中存在 Integer.MIN_VALUE 的情况
        boolean flag = true;
        int changeCount = 0;
        for (int num : nums) {
            if (num == Integer.MIN_VALUE && flag) {
                changeCount++;
                flag = false;
            }
            if (num > first) {
                changeCount++;
                third = second;
                second = first;
                first = num;
            } else if (num > second && num < first) {
                changeCount++;
                third = second;
                second = num;
            } else if (num > third && num < second) {
                changeCount++;
                third = num;
            }
        }
        return changeCount >= 3 ? third : first;
    }

}

File: python/src/problem/leetcode/easy/leetcode_414.py
# coding=utf-8
from typing import List
import sys

class Solution:
    """
    第三大的数
    """

    def third_max(self, nums: List[int]) -> int:
        """
        Time: O(n), Space: O(1)
        :param nums:
        :return:
        """
        third = second = first = -sys.maxsize
        flag = True
        change_count = 0

        for num in nums:
            if num == -sys.maxsize and flag:
                change_count += 1
                flag = False
            if num > first:
                change_count += 1
                third = second
                second = first
                first = num
            elif first > num > second:
                change_count += 1
                third = second
                second = num
            elif second > num > third:
                change_count += 1
                third = num

        return third if change_count >= 3 else first


